### ISSUE TITLE
[EMSUSD-499] Dirty layers don't get saved in Maya files

### ISSUE THREAD

**Issue by NickWu** [Autodesk/maya-usd#3243]
**Describe the bug**
Hi, I found an issue where dirty layers don't get saved in Maya files when the `mayaUsdProxyShape` is created via `stageCacheId`. 

**Steps to reproduce**

1. Run the following code in a new scene:
```python
from pxr import Usd, UsdUtils
from maya import cmds 

filePath = './proxyPrimExample.usda'
stageCache = UsdUtils.StageCache.Get()
with Usd.StageCacheContext(stageCache):
    cachedStage = Usd.Stage.Open(filePath)
    
stageId = stageCache.GetId(cachedStage).ToLongInt()
shapeNode = cmds.createNode('mayaUsdProxyShape')
cmds.setAttr('{}.stageCacheId'.format(shapeNode), stageId)
cmds.setAttr('{}.filePath'.format(shapeNode), 
    cachedStage.GetRootLayer().identifier, type='string')
```
At this point the scene looks like this:
<img width="1169" alt="Screenshot 2023-04-18 at 15 02 43" src="https://user-images.githubusercontent.com/935278/232920571-f8e38088-5e68-4e44-b25e-5aa6027f131d.png">

2. Right click on the group prim and go **Variant Sets -> geo_vis -> default** to switch its `geo_vis` variant set value from `preview` to `default`. You can see the cube now has more polygons:
<img width="1006" alt="Screenshot 2023-04-18 at 15 52 57 (1)" src="https://user-images.githubusercontent.com/935278/232922162-3ab0a335-6fc1-4b7f-8c47-2ad89720df70.png">

3. Save the Maya file with the default USD setting which is to save the layer edits (dirty layers) to the maya file:
<img width="600" alt="Screenshot 2023-07-21 at 13 59 29" src="https://github.com/Autodesk/maya-usd/assets/935278/bd4fcddc-cb3e-467f-8818-8aa079e3e624">

4. Now, close Maya and relaunch Maya. Open the saved scene, and you will see the group's `geo_vis` variant is back to its default value which is `preview`, which means the variant change we did was not saved.

<img width="600" alt="Screenshot 2023-07-21 at 14 33 31" src="https://github.com/Autodesk/maya-usd/assets/935278/a7e0fe17-db33-4408-a24c-83aaad9d4d2a">

**Expected behavior**
When the `mayaUsdProxyShape` was created via `stageCacheId`, layer edits (dirty layers) should be saved in maya files and when reopening the Maya file, we should see our changes again. 

By looking into the source code, I can see whether to save the dirty layers in maya files is determined by the following logic:

https://github.com/Autodesk/maya-usd/blob/dev/lib/mayaUsd/nodes/layerManager.cpp#L496C7-L496C7

```C++
       // Should we save the stage?
                // 1) Shareable Stage : In this case we case we only care about saving if the
                // input is not an incoming connection, since in that case the node that "owns" the
                // stage (upstream node) is responsible for saving. For example if you have multiple
                // proxy shapes daisy chained one's out_stage feeding the other's in_stage. In that
                // case only one proxy is responsible for saving (the first one).
                // 2) Unshareable Stage: Similarly but in this case if the stage is unshared, it
                // means we are responsible for saving the root layer (and stubs for the sublayers
                // so we can put them back in the same spot). So doesn't matter if its incoming or
                // not, we need to save.
                if (!pShape->isShareableStage() || !pShape->isStageIncoming()) {
                    SdfLayerHandleVector allLayers = stage->GetLayerStack(true);
                    for (auto layer : allLayers) {
                        if (layer->IsDirty()) {
                            StageSavingInfo info;
                            MDagPath::getAPathTo(mobj, info.dagPath);
                            info.stage = stage;
                            info.shareable = pShape->isShareableStage();
                            info.isIncoming = pShape->isStageIncoming();

                            // Where should we save the stage?
                            // We handle unshared composition internally in Maya USD file
                            // The reason we have this distinction now is that some Layers are
                            // special case layers that should be saved to the maya file only.
                            // There are two examples currently of this, the session layer and
                            // unshared root layer. So since we have batchSave and a delegate which
                            // handles saving externally, we need to manage some proxies ourselves
                            // and control where they save
                            if (pShape->isShareableStage()) {
                                _proxiesToSave.emplace_back(std::move(info));
                            } else {
                                _internalProxiesToSave.emplace_back(std::move(info));
                            }
                            break;
                        }
                    }
                }
```
The code comment explains the logic - if it's a sharable stage (the mayaUsdProxyShape's `shareStage` attribute is on, which it is by default), then we only save the layer changes if the mayaUsdProxyShape's input doesn't have incoming connection. If we checked the implementation of `isStageIncoming()`: https://github.com/Autodesk/maya-usd/blob/dev/lib/mayaUsd/nodes/proxyShapeBase.cpp#L1587-L1598, we can see it not only checks if the mayaUsdProxyShape's `inStageData` attribute has incoming connection but also checks whether the `stageCacheId` attribute is set and whether the value is in `UsdStageCache`. In our example, we created our mayaUsdProxyShape via `stageCacheId`, hence our changes won't get saved. In our use case, `stageCacheId` being set doesn't necessarily mean there are multiple proxy shapes sharing the same stage. Maybe we could have another attribute to force the changes to be saved when the `stageCacheId` is set? 

Also, I thought I could just turn off the `shareStage` attribute to force mayaUsdProxyShape to save the changes but then I found turning off `shareStage` attribute would cause mayaUsdProxyShape to create a new stage with a new anonymous root layer that adds the original root layer as its sub layer. This would result in two stages in the scene which is causing issues in our pipeline. We are using a shot's stage and its root layer's identifier as a key to save various pipeline data and information in our tools, but when `shareStage` attribute is off, another stage and an anonymous root layer would replace the original stage as the proxy shape's main stage, hence it's causing our tools to fail. 
 

**Attachments**
proxyPrimExample.usda
```
#usda 1.0
(
    
)

def Xform "group"
(
    prepend variantSets = "geo"
     variants = {
        string geo = "render_high"
        string geo_vis = "preview"
    }
)
{
    def Xform "GEO"
    {
    }
   variantSet "geo" = {
        "base" {
            over "GEO" (
                payload = @meshRender.usda@
            )
            {
            }

        }
        "render_high" (
            prepend variantSets = "geo_vis"
        ) {
            over "GEO" (
                payload = @meshRender.usda@
            )
            {
            }
            variantSet "geo_vis" = {
                "default" {

                }
                "preview" {
                    over "GEO"
                    {
                        rel proxyPrim = </root/GEO_PROXY>
                        uniform token purpose = "render"
                    }

                    def Xform "GEO_PROXY" (
                        payload = @meshProxy.usda@
                    )
                    {
                        uniform token purpose = "proxy"
                    }
                }
            }
        }
    }
}
```
meshProxy.usda
```
#usda 1.0
(
     defaultPrim = "root"
)
def Xform "root"
{
    def Mesh "pCube1"
    {
        float3[] extent = [(-0.5, -0.5, -0.5), (0.5, 0.5, 0.5)]
        int[] faceVertexCounts = [4, 4, 4, 4, 4, 4]
        int[] faceVertexIndices = [0, 1, 3, 2, 2, 3, 5, 4, 4, 5, 7, 6, 6, 7, 1, 0, 1, 7, 5, 3, 6, 0, 2, 4]
        point3f[] points = [(-0.5, -0.5, 0.5), (0.5, -0.5, 0.5), (-0.5, 0.5, 0.5), (0.5, 0.5, 0.5), (-0.5, 0.5, -0.5), (0.5, 0.5, -0.5), (-0.5, -0.5, -0.5), (0.5, -0.5, -0.5)]
    }
}
```
meshRender.usda
```
#usda 1.0
(
    defaultPrim = "root"
  
)

def Xform "root"
{
    def Mesh "pCube1"
    {
        float3[] extent = [(-0.5, -0.5, -0.5), (0.5, 0.5, 0.5)]
        int[] faceVertexCounts = [4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4]
        int[] faceVertexIndices = [0, 50, 74, 53, 50, 1, 51, 74, 74, 51, 4, 52, 53, 74, 52, 3, 1, 54, 75, 51, 54, 2, 55, 75, 75, 55, 5, 56, 51, 75, 56, 4, 3, 52, 76, 59, 52, 4, 57, 76, 76, 57, 7, 58, 59, 76, 58, 6, 4, 56, 77, 57, 56, 5, 60, 77, 77, 60, 8, 61, 57, 77, 61, 7, 6, 58, 78, 64, 58, 7, 62, 78, 78, 62, 10, 63, 64, 78, 63, 9, 7, 61, 79, 62, 61, 8, 65, 79, 79, 65, 11, 66, 62, 79, 66, 10, 9, 63, 80, 69, 63, 10, 67, 80, 80, 67, 13, 68, 69, 80, 68, 12, 10, 66, 81, 67, 66, 11, 70, 81, 81, 70, 14, 71, 67, 81, 71, 13, 12, 68, 82, 26, 68, 13, 72, 82, 82, 72, 16, 73, 26, 82, 73, 15, 13, 71, 83, 72, 71, 14, 27, 83, 83, 27, 17, 28, 72, 83, 28, 16, 15, 73, 84, 31, 73, 16, 29, 84, 84, 29, 19, 30, 31, 84, 30, 18, 16, 28, 85, 29, 28, 17, 32, 85, 85, 32, 20, 33, 29, 85, 33, 19, 18, 30, 86, 36, 30, 19, 34, 86, 86, 34, 22, 35, 36, 86, 35, 21, 19, 33, 87, 34, 33, 20, 37, 87, 87, 37, 23, 38, 34, 87, 38, 22, 21, 35, 88, 40, 35, 22, 39, 88, 88, 39, 1, 50, 40, 88, 50, 0, 22, 38, 89, 39, 38, 23, 41, 89, 89, 41, 2, 54, 39, 89, 54, 1, 23, 37, 90, 43, 37, 20, 32, 90, 90, 32, 17, 42, 43, 90, 42, 24, 2, 41, 91, 55, 41, 23, 43, 91, 91, 43, 24, 44, 55, 91, 44, 5, 24, 42, 92, 45, 42, 17, 27, 92, 92, 27, 14, 70, 45, 92, 70, 11, 5, 44, 93, 60, 44, 24, 45, 93, 93, 45, 11, 65, 60, 93, 65, 8, 18, 36, 94, 31, 36, 21, 46, 94, 94, 46, 25, 47, 31, 94, 47, 15, 21, 40, 95, 46, 40, 0, 53, 95, 95, 53, 3, 48, 46, 95, 48, 25, 15, 47, 96, 26, 47, 25, 49, 96, 96, 49, 9, 69, 26, 96, 69, 12, 25, 48, 97, 49, 48, 3, 59, 97, 97, 59, 6, 64, 49, 97, 64, 9]
        point3f[] points = [(-0.38888893, -0.38888893, 0.38888893), (0, -0.4375, 0.4375), (0.38888893, -0.38888893, 0.38888893), (-0.4375, 0, 0.4375), (0, 0, 0.5), (0.4375, 0, 0.4375), (-0.38888893, 0.38888893, 0.3888889), (0, 0.4375, 0.4375), (0.38888893, 0.38888893, 0.38888893), (-0.4375, 0.4375, 0), (0, 0.5, 0), (0.4375, 0.4375, 0), (-0.38888893, 0.3888889, -0.38888893), (0, 0.4375, -0.4375), (0.38888893, 0.38888893, -0.38888893), (-0.4375, 0, -0.4375), (0, 0, -0.5), (0.4375, 0, -0.4375), (-0.38888893, -0.38888893, -0.3888889), (0, -0.4375, -0.4375), (0.38888893, -0.38888893, -0.38888893), (-0.4375, -0.4375, 0), (0, -0.5, 0), (0.4375, -0.4375, 0), (0.5, 0, 0), (-0.5, 0, 0), (-0.4375, 0.25, -0.4375), (0.4375, 0.25, -0.4375), (0.25, 0, -0.5), (0, -0.25, -0.5), (-0.25, -0.4375, -0.4375), (-0.4375, -0.25, -0.4375), (0.4375, -0.25, -0.4375), (0.25, -0.4375, -0.4375), (0, -0.5, -0.25), (-0.25, -0.5, 0), (-0.4375, -0.4375, -0.25), (0.4375, -0.4375, -0.25), (0.25, -0.5, 0), (0, -0.5, 0.25), (-0.4375, -0.4375, 0.25), (0.4375, -0.4375, 0.25), (0.5, 0, -0.25), (0.5, -0.25, 0), (0.5, 0, 0.25), (0.5, 0.25, 0), (-0.5, -0.25, 0), (-0.5, 0, -0.25), (-0.5, 0, 0.25), (-0.5, 0.25, 0), (-0.25, -0.4375, 0.4375), (0, -0.25, 0.5), (-0.25, 0, 0.5), (-0.4375, -0.25, 0.4375), (0.25, -0.4375, 0.4375), (0.4375, -0.25, 0.4375), (0.25, 0, 0.5), (0, 0.25, 0.5), (-0.25, 0.4375, 0.4375), (-0.4375, 0.25, 0.4375), (0.4375, 0.25, 0.4375), (0.25, 0.4375, 0.4375), (0, 0.5, 0.25), (-0.25, 0.5, 0), (-0.4375, 0.4375, 0.25), (0.4375, 0.4375, 0.25), (0.25, 0.5, 0), (0, 0.5, -0.25), (-0.25, 0.4375, -0.4375), (-0.4375, 0.4375, -0.25), (0.4375, 0.4375, -0.25), (0.25, 0.4375, -0.4375), (0, 0.25, -0.5), (-0.25, 0, -0.5), (-0.25, -0.25, 0.5), (0.25, -0.25, 0.5), (-0.25, 0.25, 0.5), (0.25, 0.25, 0.5), (-0.25, 0.5, 0.25), (0.25, 0.5, 0.25), (-0.25, 0.5, -0.25), (0.25, 0.5, -0.25), (-0.25, 0.25, -0.5), (0.25, 0.25, -0.5), (-0.25, -0.25, -0.5), (0.25, -0.25, -0.5), (-0.25, -0.5, -0.25), (0.25, -0.5, -0.25), (-0.25, -0.5, 0.25), (0.25, -0.5, 0.25), (0.5, -0.25, -0.25), (0.5, -0.25, 0.25), (0.5, 0.25, -0.25), (0.5, 0.25, 0.25), (-0.5, -0.25, -0.25), (-0.5, -0.25, 0.25), (-0.5, 0.25, -0.25), (-0.5, 0.25, 0.25)]
    }
}
```

**Specs (if applicable):**
 - OS & version:  CentOS 7.8
 - Compiler & version: GCC 9.3.1
 - Maya version: Maya-2023.2.0-18.1.PFIX
 - Maya USD commit SHA: https://github.com/Autodesk/maya-usd/commit/c455f25caf420057b556b2e28c9154bbc5cff009

**Comment by santosd** [comment]
Logged this internally so the team can take a look. EMSUSD-499.

**Comment by frenchdog** [comment]
Hi @NickWu,
Indeed, when a mayaUSDProxyShape uses a stage from the stage cache, its layers are never saved. This is by design for reasons I'm going to explain bellow, but first I would like to say that in your scenario, I think it might be sensible to let Maya save layers used by the stage cache.
So the current design assume that an upstream Maya node can create a stage and is able to save such stage by its own. 
A real example is Bifrost. A Bifrost graph can create stages and can save them to disk too. In order to display its stage in Maya viewport, Bifrost uses the stage cache too. On the Bifrost side, the stage is added to the cache and its id is connected to the "stageCacheId" plug of a mayaUsdProxyShape. In such case, when saving the Maya scene, the layers of the Bifrost stage should never be saved in the Maya scene since Bifrost could potentially generate a lot of USD data (and such data would be generated by Bifrost when re-opening the Maya scene anyway). 
But in your scenario, since you are creating a stage and setting the "stageCacheId" in Python (without creating any connection between nodes), I think we should save the content of the "cached stage" in the Maya file.
In short the logic could be:
If "mayaUsdProxyShape.stageCacheId" is connected, don't save, else if the stageCacheId refer an existing stage save its content.

**Comment by pierrebai-adsk** [comment]
What Guillaume just said makes sense and if a PR containing just that modification were open, we would probably merge it in.

**Comment by NickWu** [comment]
Hi @frenchdog , thank you for your detail explanation. Yes, the logic you described 

> If "mayaUsdProxyShape.stageCacheId" is connected, don't save, else if the stageCacheId refer an existing stage save its content.

was exactly what I had in mind.

**Comment by NickWu** [comment]
I'll submit a PR in the coming weeks for this.

**Comment by NickWu** [comment]
I have some feedback on the PR https://github.com/Autodesk/maya-usd/pull/3402#discussion_r1369448291

**Comment by santosg87** [comment]
@NickWu can this issue be closed now?

**Comment by santosd** [comment]
Validated the changes on my end and it seems to be working as expected. Thank you @NickWu for working with @samuelliu-adsk on this. I will go ahead and close this ticket out since the issue has been resolved.

**Comment by NickWu** [comment]
Thank you.
